src/lazy/array.ts

```diff
@@ -0,0 +1,200 @@
+import { AnyNode } from "../traverse";
+import {
+	ILazyArray,
+	MapFn,
+	FilterFn,
+	FlatMapFn,
+	SeedIterable,
+	EachFn,
+	Type,
+	DeepCollected,
+	FlattenedArray,
+	ReduceFn,
+	ReduceFnRT,
+	ObjectKey,
+} from "./abstract";
+import hasType from "./hasType";
+import { getEntries, rCollectDeep } from "./iterable";
+import iterator from "./iterator";
+import { LazyObject } from "./object";
+import { LazySet } from "./set";
+import { LazyValue } from "./value";
+
+export class LazyArray<T> implements ILazyArray<T> {
+	__type: Type.Array = Type.Array;
+	constructor(private readonly seed: SeedIterable<T, number>) {}
+	collect(): DeepCollected<T, Type.Array> {
+		return [...this].map((item) => rCollectDeep(item as unknown as AnyNode)) as DeepCollected<T, Type.Array>;
+	}
+	iterators: { values(): IterableIterator<T>; indices(): IterableIterator<number>; entries(): IterableIterator<[number, T]> } = (() => {
+		const getIterable = () => getEntries(this.seed);
+		return {
+			*values() {
+				for (const [, value] of this.entries()) {
+					yield value;
+				}
+			},
+			*indices() {
+				for (const [index] of this.entries()) {
+					yield index;
+				}
+			},
+			*entries() {
+				yield* getIterable();
+			},
+		};
+	})();
+	map<U>(f: MapFn<T, number, U>): ILazyArray<U> {
+		return new MapIterator(this, f);
+	}
+	*each(f: EachFn<T, number>): Generator {
+		for (const [index, value] of this.iterators.entries()) {
+			yield f(value, index);
+		}
+	}
+	filter(f: FilterFn<T, number>): ILazyArray<T> {
+		return new FilterIterator(this, f);
+	}
+	flatten(): FlattenedArray<T> {
+		return new FlattenIterator(this) as unknown as FlattenedArray<T>;
+	}
+	flatMap<U>(f: FlatMapFn<T, number, U>): ILazyArray<U> {
+		return this.map(f).flatten() as ILazyArray<U>;
+	}
+	concat(iterable: SeedIterable<T, number>): ILazyArray<T> {
+		return new ConcatIterator(this, iterable);
+	}
+	append(item: T): ILazyArray<T> {
+		return this.concat([item]);
+	}
```

Comment:
I hate to make this comment but, just for readability, could we add some empty lines after each method.

Reply:
Piece of cake